### ISSUE TITLE
Cannot Get UWP App Using Microsoft.Data.SqlClient to Pass Certification for Microsoft Store

### ISSUE THREAD

**Issue by wjvii** [dotnet/SqlClient#703]
Trying to get our application certified for the store and we keep getting it rejected with the following:

API IsTokenRestricted in advapi32.dll is not supported for this application type. Microsoft.Data.SqlClient.SNI.dll calls this API.
API CertOpenSystemStoreA in crypt32.dll is not supported for this application type. Microsoft.Data.SqlClient.SNI.dll calls this API.
API CryptStringToBinaryW in crypt32.dll is not supported for this application type. Microsoft.Data.SqlClient.SNI.dll calls this API.

I am not sure how to get past this and what is odd is that we have .NET apps that use it that we package up and they are not rejected in the store certification process. This application we cannot get to pass was created as UWP and we are not sure how to proceed to get it through the store. Is this a feature request for Microsoft.Data.SqlClient, a store certification request. At this point just wondering if anyone else has run into this and if anyone has guidance on using this in a UWP application.

Thank you in advance.

**Comment by JRahnama** [comment]
@wjvii thanks for bringing up the issue here, if I am not wrong the [issue#356](https://github.com/dotnet/SqlClient/issues/356) is about the same. Can you tell us what version of Microsoft.Data.SqlClient you are using?

**Comment by wjvii** [comment]
@JRahnama thank you for the super quick response. It is Microsoft.Data.SqlClient 2.0.0 which appears to pull in Microsoft.Data.SqlClient.SNI.runtime. I see in the other thread that some work was done with System.Data.SqlClient to make it work with UWP and this solution still has a component that uses System.Data.SqlClient and as stated it does not appear to be a problem in the certification process.

**Comment by cheenamalhotra** [comment]
Hi @wjvii 

System.Data.SqlClient shipped UAP DLLs which had Managed SNI enabled on Windows.

With Microsoft.Data.SqlClient, we don't have specific DLLs for UAP so .NET Standard DLLs are downloaded.
But Managed SNI on Windows is disabled by default, which can be enabled by setting context switch as documented in [Microsoft Docs](https://docs.microsoft.com/en-us/sql/connect/ado-net/introduction-microsoft-data-sqlclient-namespace?view=sql-server-ver15#enabling-managed-networking-on-windows).

```cs
AppContext.SetSwitch("Switch.Microsoft.Data.SqlClient.UseManagedNetworkingOnWindows", true);
```

This will bring equivalent behavior in UWP applications as System.Data.SqlClient.

**Comment by wjvii** [comment]
@cheenamalhotra thank you so much for the information. Where do I add that AppContext.SetSwitch statement, just any time before I use the Microsoft.Data.SqlClient? So, if I got you right doing this will make it pass certification. That is awesome. I can suggest the certification team add this to their report as a suggestion when it fails related to Microsoft.Data.SqlClient.

**Comment by cheenamalhotra** [comment]
Yes, you can enable with context switch to any startup path/controller, where it's feasible or in configuration file.
More info: [AppContext for library consumers](https://docs.microsoft.com/en-us/dotnet/api/system.appcontext?view=netcore-3.1#appcontext-for-library-consumers), [AppContext.SetSwitch](https://docs.microsoft.com/en-us/dotnet/api/system.appcontext.setswitch?view=netcore-3.1#System_AppContext_SetSwitch_System_String_System_Boolean_)

**Comment by cheenamalhotra** [comment]
Closing this issue as UWP Apps require enabled App Context switch to use Managed SNI on Windows.

**Comment by wjvii** [comment]
Ran into a similar issue again and in this case it was because Entity Framework Core 3.1 has a dependency on Microsoft.Data.SqlClient 1.1.3 and enabling Managed SNI on Windows via the AppContext switch was not available until Microsoft.Data.SqlClient 2.0. Added the Microsoft.Data.SqlClient 2.0 NuGet to the project so it would include the 2.0+ version and it resolved the issue with the Microsoft App Store when used along with the AppContext switch.

**Comment by wjvii** [comment]
Ran into the same issue again and in this case it was because Entity Framework Core 3.1 has a depdency on Microsoft.Data.SqlClient 1.1.3 and enabling Managed SNI on Windows via the AppContext switch wasn't available until Microsoft.Data.SqlClient 2.0 or later. Added a direct reference to

**Comment by mvegaca** [comment]
Hi

I'm working on Windows Template Studio [Issue 3951](https://github.com/microsoft/WindowsTemplateStudio/issues/3951) trying to update Microsoft.EntityFrameworkCore to 3.1.10 version.

I've tried to add the AppContext Switch and also added the nuget Microsoft.Data.SqlClient version 2.0.0.

```csharp
// Added on ActivationService class, InitializeAsync method
AppContext.SetSwitch("Switch.Microsoft.Data.SqlClient.UseManagedNetworkingOnWindows", true);
```

But I keep getting the same error when executing WACK tests.

```
Supported APIs
Error Found: The supported APIs test detected the following errors:
API ExecuteAssembly in uwphost.dll is not supported for this application type. EmulatorApp2.exe calls this API.
API DllGetActivationFactory in uwphost.dll is not supported for this application type. EmulatorApp2.exe has an export that forwards to this API.
API IsTokenRestricted in advapi32.dll is not supported for this application type. Microsoft.Data.SqlClient.SNI.dll calls this API.
API CertOpenSystemStoreA in crypt32.dll is not supported for this application type. Microsoft.Data.SqlClient.SNI.dll calls this API.
API CryptStringToBinaryW in crypt32.dll is not supported for this application type. Microsoft.Data.SqlClient.SNI.dll calls this API.
Impact if not fixed: Using an API that is not part of the Windows SDK for Microsoft Store apps violates the Microsoft Store certification requirements.
How to fix: Review the error messages to identify the API that is not part of the Windows SDK for Microsoft Store apps. Please note, apps that are built in a debug configuration or without .NET Native enabled (where applicable) can fail this test as these environments may pull in unsupported APIs. Retest your app in a release configuration, and with .NET Native enabled if applicable. See the link below for more information:
```
![image](https://user-images.githubusercontent.com/5556784/100092765-e50f9180-2e56-11eb-8a5e-77700e61c3f4.png)

This is a repro app created for this proposal
[EmulatorApp2.zip](https://github.com/dotnet/SqlClient/files/5589751/EmulatorApp2.zip)

Am I missing something?

**Comment by cheenamalhotra** [comment]
Hi @mvegaca 

You're getting these errors as SNI DLL is still loaded in process, as they're referenced by SqlClient. Even if you enable App Context switch, the driver will not use the native libraries but they're still loaded.

Is it possible to ignore these errors and test run your application?

**Comment by mvegaca** [comment]
Hi @cheenamalhotra, thanks for your answer. Windows Template Studio's goal is to help devs bootstrap UWP applications. We need the WACK tests to pass for the code we generate. We'll stick with Microsoft.EntityFrameworkCore.SqlServer Version 2.2.6 for now until this can be fixed.

**Comment by wjvii** [comment]
I was once again greeted with this same issue trying to get our application certified in the Store after updating to EF Core 3.1.9. Is there anything that we can do so applications that use Entity Framework and Microsoft.Data.SqlClient can pass certification in the Microsoft Store using EF Core 3.0 or later? We already directly reference Microsoft.Data.SqlClient 2.0 and we already have the AppContext.SetSwitch("Switch.Microsoft.Data.SqlClient.UseManagedNetworkingOnWindows", true); 

If it is not possible to make the package build without these, could you work with the Microsoft Store team and get them to allow this?

Thank you in advance.

**Comment by 3AM-Innovations** [comment]
I have an app using EF Core 3.1.4 and Microsoft Data Sql Client 2.1.2 that I upload just 10 days ago that has the reference to the Microsoft.Data.SqlClient.SNI.dll in the appx bundle. It's literally available in the store today. However the latest version that I was trying to upload just yesterday is getting rejected due to the Microsoft.Data.SqlClient.SNI.dll file. 

What I don't understand is why was the current version accepted but the latest build not? Both have the file...

**Comment by Tuytje** [comment]
I'm running into the same issue with a Xamarin Forms app & adding the Microsoft.Data.SqlClient library + Context switch doesn't help. It seems that every time I need to build a new version for the Windows Store, something new pops up :/

**Comment by kaki104** [comment]
Hi
I am going to make a UWP app with the Uno Platform and register it in the Store. 
The first error message I got was below: 

API IsTokenRestricted in advapi32.dll is not supported for this application type. sni.dll calls this API

I came here looking for a way to fix that error. 

Put the following in the App constructor, 
AppContext.SetSwitch("Switch.Microsoft.Data.SqlClient.UseManagedNetworkingOnWindows", true);
Added Microsoft.Data.SqlClient version 2.1.2 Nuget package. 

The final WACK result is as follows. 

API IsTokenRestricted in advapi32.dll is not supported for this application type. Microsoft.Data.SqlClient.SNI.dll calls this API.
API CertOpenSystemStoreA in crypt32.dll is not supported for this application type. Microsoft.Data.SqlClient.SNI.dll calls this API.
API CryptStringToBinaryW in crypt32.dll is not supported for this application type. Microsoft.Data.SqlClient.SNI.dll calls this API.

What should I do now? please tell me how can i fix this error T_T 
@cheenamalhotra

**Comment by cheenamalhotra** [comment]
Hi @kaki104 

Is it possible to remove the Microsoft.Data.SqlClient.SNI.dll from your application package before deployment? You are safe to do so if you've enabled the App Context switch. Please explore that idea.

**Comment by kaki104** [comment]
@cheenamalhotra 
I found information on how to delete the Microsoft.Data.SqlClient.SNI.dll file. However, that method is difficult to use. I tried, but it failed. 
In the end, I removed the Microsoft.EntityFrameworkCore.SqlServer nuget package from the Repository project used by the UWP app. So, by removing the dependency between the UWP app and Microsoft.Data.SqlClient, it passed WACK.
This time, I solved it by modifying the source, but I hope other developers can use the same method to solve it later.
